### PR TITLE
[ENH] Add range_flag function

### LINKED ISSUE #708: Add the ability to test whether data is in range in a pandas data frame
# Brief Description

Following up on #703, this issue seeks to introduce the ability to test whether each column in a data frame is within a user-defined range of values. 

I would like to propose..

# Example API

```python
def flag_values_not_in_range(
        df: pd.DataFrame,
        bound: list,
        inclusive: bool = True
) -> pd.DataFrame:
    """
    :param df: data frame to test if values in columns are within a range
    :param bound: user-defined lower and upper bound of values to test in data frame columns
    :param inclusive: include boundaries in filtering.
        Defaults to True
    :return: True if value in column does not fall within range.
        False if value in column falls within range.
    """

```

### PR DESCRIPTION
**Pull request by UGuntupalli**
#708  

PR Description

Please describe the changes proposed in the pull request:

- Added a function that flags whether data in columns are within a user defined range.

**This PR resolves #708 .**

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.
2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.
3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by UGuntupalli** [comment]
One thing I am thinking as I make the submission is if it would be more beneficial to make a ```bound``` a dictionary of lists, so a different set of limits can be passed for each column -  Just a thought, but did not want to over complicate it

**Comment by ericmjl** [comment]
@UGuntupalli thanks for the PR! As I looked through the code, I realized that this behaves as a "data validation function", rather than a "data cleaning function", and was wondering if this is better done as a `pandera` check, rather than a `pyjanitor` function, maybe? For example, if you have a dataframe for which you need to _validate_ that the values are within a given range, a possibility is as follows:

```python
import pandera as pa 

schema = pa.DataFrameSchema(
    columns={
        "column_to_check": pa.Column(pa.DateTime, checks=pa.Check.in_range(...),
    }
)
```

And then you would simply flag error using:

```python
schema.validate(df, lazy=True)
```

Immediately, an error would show up, _and_ `pandera` would also give you the exact rows that fail the check.

The use of `pandera` is something I definitely recommend for data validation, as it allows us to very easily declare up-front what we expect from our data, and catch the errors at runtime.

What are your thoughts here?

Apart from the core of the proposed function possibly being better done in a different package, I must say, your PR was quite well done, and I'm super happy to see the improvement! :smile: :heart:

File: janitor/timeseries.py [hunk 1]
```diff
@@ -87,3 +87,54 @@ def _get_missing_timestamps(
...
+    :param inclusive: flag to include boundaries
+    :returns: dataframe with booleans for each column
+    """
+    # Check the input data sets for data type
+    check("df", df, [pd.DataFrame])
+    check("bound", bound, [list])
+    check("inclusive", inclusive, [bool])
+
+    # Get names of the columns into a list
+    columns = df.columns
```
Comment (samukweku) [inline janitor/timeseries.py:126]: You can reduce the number of lines and avoid the for loop by using pandas transform, then assigning the result to the new dataframe : 
```python
df1 = df.transform(pd.Series.between, axis=0, left=bound[0], right=bound[-1])
```
You'll have to test the speed though; this however is a built-in pandas function and negates the need for the for loop and the temporary variables for columns.

Also, there is a [between_time](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.between_time.html#:~:text=between_time,-DataFrame.&text=Select%20values%20between%20particular%20times,not%20between%20the%20two%20times.) function as well, with left and right bounds. It could possibly be a better option, since it is dedicated to time manipulations exclusively.

**Review by samukweku — CHANGES_REQUESTED** [review]

File: janitor/timeseries.py (hunk 1 above)
Comment (UGuntupalli) [reply janitor/timeseries.py:126]: @samukweku , 
           I like the idea of keeping it native to Pandas, so I am happy to make this modification, remove the loop and replace it with ```transform()```. The ```between_time``` is only to handle datetimes correct- in my case the data can be anything and not necessarily datetimes. 

@ericmjl,
          Just to make sure I understand your suggestion correctly: 

1. Are you suggesting implementing the function in ```pandera``` package ? 
2. Are you suggesting re-writing the function using ```pandera``` in ```pyjanitor```? 

If it is (2), I believe ```pyjanitor```development environment does not have ```pandera```included. If it is (1), I can get started trying to get familiar with the process of contributing to ```pandera``` . One issue I see with (1) is, we don't necessarily need an error to be generated, rather just generate a mask of booleans, I am not sure if that is possible with the ```check``` construct in ```pandera```. This is something I will have to check

Comment (ericmjl) [reply janitor/timeseries.py:126]: @UGuntupalli I took a few days to think more carefully about this PR, and I think I was previously getting confused by the PR name. "Testing" whether a column's values are inside a pre-defined range is a different activity from "flagging" whether that column's values are inside a pre-defined range. The difference lies in what actions are taken: a test should error out if the dataframe column fails a check, while a flag will add a column that contains the flag. Tests are out-of-scope for `pyjanitor`, because there's better tooling for this (i.e. `pandera`), but flagging is not, because we already have precedent (`flag_nulls`). Imprecise semantics were what were confusing me before.

The function name here contains `flag`, which puts it under the flagging category of things. I've renamed the PR and edited the description accordingly. Let's proceed with the PR; you can also ignore my comment about `pandera`, which I typed under the assumption that the intent was to put in a data testing function. (Data tests should be contributed to `pandera` if they aren't available in there already.)

File: janitor/timeseries.py [hunk 2]
```diff
@@ -87,3 +87,54 @@ def _get_missing_timestamps(
     missing_timestamps = expected_df.index.difference(df.index)
 
     return expected_df.loc[missing_timestamps]
+
+
+@pf.register_dataframe_method
+def range_flag(
```
Comment (ericmjl) [inline janitor/timeseries.py:93]: The behaviour of this function seems to be to return a completely different dataframe from the original, rather than appending new columns to the original.

This deviates from the behaviour of _most_ of the functions in `pyjanitor`, and I think the behaviour has to be properly documented in the docstring.

Question for you, @UGuntupalli - is the intent here to add columns of flags to the dataframe, or to return a completely new dataframe that has the flags, i.e. with the behaviour of `pandas`-native `df.transform()`?

I'd prefer that the behaviour would match `flag_nulls`, by adding a column of booleans to the dataframe, rather than following `df.transform()`, because this would make the behaviour of `flag_` functions consistent across the library. (I'd also rename the function to `flag_in_range`.) If you want to make the function use the latter behaviour (i.e. new df with boolean flags), I'd name the function differently so that the function-name-to-behaviour consistency is maintained. My naming proposals in the latter case would be:

- `isin_range(...)` - has the same behaviour as `df["column"].isin(some_list)`. I would favour this name, as it conveys the same behaviour that other users would expect.
- `highlight_in_range(...)` - thus kickstarting a class of functions called `highlight_X`

Sorry if it feels like I'm being a bit picky with the naming, but as the library grows, new additions have to fit into the right category of things in order to ensure that the behaviour of the library is consistent.

File: janitor/timeseries.py [hunk 3]
```diff
@@ -87,3 +87,54 @@ def _get_missing_timestamps(
...
+    check("inclusive", inclusive, [bool])
+
+    # Get names of the columns into a list
+    columns = df.columns
+
+    # Make a copy to avoid mutation of the source dataframe
+    df1 = pd.DataFrame(index=df.index)
+
+    # Loop through the columns and create new flags
+    for column in columns:
```
Comment (ericmjl) [inline janitor/timeseries.py:132]: I'd recommend using the `df.assign` pattern here:

```python
columns_data = {
    f"{column}_range_flag": df[column].between(left_bound=bound[0], right_bound=bound[1], inclusive=inclusive]
    for column
    in df.columns
}
df = df.assign(**columns_data)
```

This seems to be the more `pandas`-native way to assign new columns to the dataframe. The older selector-style syntax modifies the original. I think it's an opportunity here to encourage other readers of the source code to use this style; this is a change I want to make across the library.

File: tests/timeseries/test_range_flag.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@pytest.mark.timeseries
+def test_range_flag_inclusive(timeseries_dataframe):
+    """ Test range flag with inclusive argument """
+    # Set some values equal to boundary values
+    random_low = randint(1, len(timeseries_dataframe))
+    random_high = randint(1, len(timeseries_dataframe))
+    timeseries_dataframe.loc[timeseries_dataframe.index[random_low]] = 100
+    timeseries_dataframe.loc[timeseries_dataframe.index[random_high]] = 750
```
Comment (ericmjl) [inline tests/timeseries/test_range_flag.py:48]: I see the magic number 100 and 750 show up quite a bit. It'd be helpful for further test development later if you could assign them to a variable inside the function and use the variables instead. That way, if someone else has the bandwidth and wants to parametrize the test, they'd be able to do so.

File: tests/timeseries/test_range_flag.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Test if bounds have been honored
+    v1_subset_minimum = v1_subset.min()
+    v1_subset_maximum = v1_subset.max()
+    assert(100 == v1_subset_minimum)
+    assert(750 == v1_subset_maximum)
+
+
+if __name__ == '__main__':
+    pytest.main()
```
Comment (ericmjl) [inline tests/timeseries/test_range_flag.py:70]: I think we don't use this pattern elsewhere in the test suite. Where did you read about this?

**Review by ericmjl — CHANGES_REQUESTED** [review]

**Comment by ericmjl** [comment]
Hey @UGuntupalli! Just wanted to check in, anything I can do to help move this PR along? No pressure here, I'm working under the assumption that you've got your own open source projects that you'd like to get kickstarted too :).

**Comment by UGuntupalli** [comment]
@ericmjl , 
            Just slammed both personally and professionally. It might be about a week or so before I am able to pick this up again unfortunately :(. I definitely want to finish this and contribute more inspite of the other open source project we discussed, but am a little behind with things. If anybody is able to adopt the changes suggested in the PR review - then great, else I will finish it in a weeks time

**Comment by samukweku** [comment]
@UGuntupalli, just leaving a note here that using pd.Interval may be faster/more efficient

**Comment by ericmjl** [comment]
Going to close this PR because it is quite outdated and the codebase has moved on. Leaving for the record that I do value the contributions regardless!
